lib/std/crypto/scrypt/scrypt.zig

```diff
@@ -0,0 +1,244 @@
+        blockMix(&tmp, y, x, r);
+    }
+
+    for (x) |v1, j| {
+        mem.writeIntLittle(u32, b[4 * j ..][0..4], v1);
+    }
+}
+
+const Error1 = error{
+    InvalidParams,
+    InvalidDerivedKeyLen,
+};
+
+// +Pbkdf2Error
+pub const Error = Error1 || mem.Allocator.Error;
+
+pub const Params = struct {
+    const Self = @This();
+
+    log_n: u6 = 15,
+    r: u32 = 8,
+    p: u32 = 1,
+
+    pub fn init(log_n: u6, r: u32, p: u32) Self {
+        return Self{ .log_n = log_n, .r = r, .p = p };
+    }
+
+    pub fn fromLimits(ops_limit: u64, mem_limit: usize) Self {
+        const ops = math.max(32768, ops_limit);
+        const r: u32 = 8;
+        if (ops < mem_limit / 32) {
+            const max_n = ops / (r * 4);
+            return Self{ .r = r, .p = 1, .log_n = @intCast(u6, math.log2(max_n)) };
+        } else {
+            const max_n = mem_limit / (@intCast(usize, r) * 128);
+            const log_n = @intCast(u6, math.log2(max_n));
+            const max_rp = math.min(0x3fffffff, (ops / 4) / (@as(u64, 1) << log_n));
+            return Self{ .r = r, .p = @intCast(u32, max_rp / @as(u64, r)), .log_n = log_n };
+        }
+    }
+
+    pub fn fromPhcString(s: []const u8) phc.Error!Self {
+        var res = Self{};
+        var it = phc.ParamsIterator.init(s, 3);
+        while (try it.next()) |param| {
+            if (mem.eql(u8, param.key, "ln")) {
+                res.log_n = try param.decimal(u6);
+            } else if (mem.eql(u8, param.key, "r")) {
+                res.r = try param.decimal(u32);
+            } else if (mem.eql(u8, param.key, "p")) {
+                res.p = try param.decimal(u32);
+            } else {
+                return error.ParseError;
+            }
+        }
+        return res;
+    }
+
+    pub fn toPhcString(self: Self, allocator: *mem.Allocator) mem.Allocator.Error![]const u8 {
+        const buf = try fmt.allocPrint(
+            allocator,
+            "ln={d},r={d},p={d}",
+            .{ self.log_n, self.r, self.p },
+        );
+        return buf;
+    }
+};
+
+// TODO return Error
+
+/// Apply SCRYPT to generate a key from a password.
+///
+/// SCRYPT is defined in RFC 7914.
+///
+/// allocator: *mem.Allocator.
+///
+/// derived_key: Slice of appropriate size for generated key. Generally 16 or 32 bytes in length.
+///              May be uninitialized. All bytes will be overwritten.
+///              Maximum size is `derived_key.len / 32 == 0xffff_ffff`.
+///              It is a programming error to pass buffer longer than the maximum size.
+///
+/// password: Arbitrary sequence of bytes of any length.
+///
+/// salt: Arbitrary sequence of bytes of any length.
+///
+/// params: Optional Params. Defaults may change in future.
+pub fn kdf(
+    allocator: *mem.Allocator,
+    derived_key: []u8,
+    password: []const u8,
+    salt: []const u8,
+    params: ?Params,
+) !void {
+    if (derived_key.len == 0 or derived_key.len / 32 > 0xffff_ffff) {
+        return error.InvalidDerivedKeyLen;
```

Comment:
@daurnimator  Thats a bad design choice. Why avoid using good error handling system? It should not reduce performance as much as exceptions would and it allows to use these apis on non verified input. I see no reasons to prefer DoS.

Reply:
The output length is not something controlled by an attacker. Or if it is, the DoS vector exists without an `assert()`.

Reply:
@jedisct1 The problem with this is that it encourages external parameter validation, duplicating code, and possibly introducing bugs, let the api report the error to me and will decide what to do, use `catch unreachable` or log or warn the user. Especially since there is `error.OutputTooLong` in crypto/error.zig!

Especially i dont like `assert(msg.len == out.len)` (ive seen this is crypto/salsa20.zig), this probably a bug that it wont let me overallocate output slice. But anyway i would prefer `error.OutputOverflow` instead of `assert(out.len >= msg.len)`.

Reply:
I agree with @Rocknest here. Using asserts is for things that are expected. If `dk.len == 0` maybe is expected, than `dk.len / 32 > 0xffff_ffff` is not expected at all. Somebody 100% will "shoot its leg".